numba/cuda/cudadrv/devicearray.py

```diff
@@ -618,20 +626,34 @@ def from_record_like(rec, stream=0, gpu_data=None):
     return DeviceRecord(rec.dtype, stream=stream, gpu_data=gpu_data)
 
 
+def array_core(ary):
```

Comment:
No. Consider the following example:
```python
>>> import numpy as np
>>> from numba import cuda      
>>> array_core = cuda.cudadrv.devicearray.array_core
>>> full = np.broadcast_to(np.arange(6).reshape((3, 2, 1)), (3, 2, 2))[:, :1, :]
>>> full
array([[[0, 0]],

       [[2, 2]],

       [[4, 4]]])
>>> full.size, full.shape, full.strides
(6, (3, 1, 2), (16, 8, 0))
>>> squeezed = full.squeeze()
>>> squeezed
array([[0, 0],
       [2, 2],
       [4, 4]])
>>> squeezed.size, squeezed.shape, squeezed.strides
(6, (3, 2), (16, 0))
>>> core = array_core(full)
>>> core
array([[0],
       [2],
       [4]])
>>> core.size, core.shape, core.strides
(3, (3, 1), (16, 8))
```
As this demonstrates, `squeeze` and `array_core` are orthogonal: `squeeze` removes axes of *length* one and arbitrary stride; such axes can always be added to or removed from the shape without changing the *logical* size and content of the array, for example using `reshape`. On the other hand, `array_core` removes axes of *stride* zero and arbitrary length; such axes can always be added to or removed from the array without changing the *physical* size and content of the array in memory, for example using `broadcast_to`, but will change the logical size and content by repeating a "core" array. We need to access this core in order to correctly copy data between such arrays.

If you know a more elegant way of accessing this core, let me know! I couldn't find any, so I wrote this function. Also, let me know if you'd rather have the function in a general utils module rather than within the cuda subpackage; it works for any strided array type with numpy-like indexing and attributes. The same also goes for the `check_array_compatibility` function.